docs/cloud-integration.md

```diff
@@ -0,0 +1,953 @@
+The steps to do so are as follows
+
+1. Use the full URI to refer to a bucket, including the prefix for the client-side library
+to use. Example: `s3a://landsat-pds/scene_list.gz`
+1. Have the Spark context configured with the authentication details of the object store.
+In a YARN cluster, this may also be done in the `core-site.xml` file.
+1. Have the JAR containing the filesystem classes on the classpath —along with all of its dependencies.
+
+### <a name="dataframes"></a>Example: DataFrames
+
+DataFrames can be created from and saved to object stores through the `read()` and `write()` methods.
+
+{% highlight scala %}
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.types.StringType
+
+val spark = SparkSession
+    .builder
+    .appName("DataFrames")
+    .config(sparkConf)
+    .getOrCreate()
+import spark.implicits._
+val numRows = 1000
+
+// generate test data
+val sourceData = spark.range(0, numRows).select($"id".as("l"), $"id".cast(StringType).as("s"))
+
+// define the destination
+val dest = "wasb://yourcontainer@youraccount.blob.core.windows.net/dataframes"
+
+// write the data
+val orcFile = dest + "/data.orc"
+sourceData.write.format("orc").save(orcFile)
+
+// now read it back
+val orcData = spark.read.format("orc").load(orcFile)
+
+// finally, write the data as Parquet
+orcData.write.format("parquet").save(dest + "/data.parquet")
+spark.stop()
+{% endhighlight %}
+
+### <a name="streaming"></a>Example: Spark Streaming and Cloud Storage
+
+Spark Streaming can monitor files added to object stores, by
+creating a `FileInputDStream` DStream monitoring a path under a bucket.
+
+{% highlight scala %}
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.streaming._
+
+val sparkConf = new SparkConf()
+val ssc = new StreamingContext(sparkConf, Milliseconds(5000))
+try {
+  val lines = ssc.textFileStream("s3a://bucket/incoming")
+  val matches = lines.filter(_.endsWith("3"))
+  matches.print()
+  ssc.start()
+  ssc.awaitTermination()
+} finally {
+  ssc.stop(true)
+}
+{% endhighlight %}
+
+1. The time to scan for new files is proportional to the number of files
+under the path —not the number of *new* files, and that it can become a slow operation.
+The size of the window needs to be set to handle this.
+
+1. Files only appear in an object store once they are completely written; there
+is no need for a worklow of write-then-rename to ensure that files aren't picked up
+while they are still being written. Applications can write straight to the monitored directory.
+
+#### <a name="checkpointing"></a>Checkpointing Streams to object stores
+
+Streams should only be checkpointed to an object store considered compatible with
+HDFS. As the checkpoint operation includes a `rename()` operation, checkpointing to
+an object store can be so slow that streaming throughput collapses.
+
+
+## <a name="output"></a>Object Stores as a substitute for HDFS
```

Comment:
It really depends on what you want to do. It works against Azure because that's consistent enough. It can work with EMR S3 using DynamoDB for consistency, it will work once we do the same in ASF Hadoop with s3guard (if you make dynamo authoritative). You can use s3 data as a source of information, especially stuff that doesn't change very often. Using it as a in intermediate step between queries is not likely to work right now